scalyr_agent/tests/copying_manager_test.py

```diff
@@ -784,6 +821,186 @@ def test_pipelined_requests_with_retry_error(self):
 
         responder_callback('success')
 
+    def test_start_from_full_checkpoint(self):
+        controller = self.__create_test_instance()
+        previous_root_dir = os.path.dirname(self.__test_log_file)
+
+        self.__append_log_lines('First line', 'Second line')
+        (request, responder_callback) = controller.wait_for_rpc()
+        lines = self.__extract_lines(request)
+
+        self.assertEquals(2, len(lines))
+        self.assertEquals('First line', lines[0])
+        self.assertEquals('Second line', lines[1])
+
+        # stop thread on manager to write checkouts to file.
+        controller.stop()
+
+        # write some new lines to log.
+        self.__append_log_lines('Third line', 'Fourth line')
+
+        # create new copying manager, but passing previous directory with same log and checkouts.
+        controller = self.__create_test_instance(root_dir=previous_root_dir, auto_start=False)
+
+        self._manager.start_manager(dict(fake_client=True))
+        (request, responder_callback) = controller.wait_for_rpc()
+        lines = self.__extract_lines(request)
+
+        # thread should continue from saved checkpoint
+        self.assertEquals(2, len(lines))
+        self.assertEquals('Third line', lines[0])
+        self.assertEquals('Fourth line', lines[1])
+
+        # stopping one more time, but now emulating that checkpoint files are stale.
+        controller.stop()
+
+        self.__append_log_lines('Fifth line', 'Sixth line')
+
+        def shift_time_in_checkpoint_file(name):
+
+            fp = open(os.path.join(self._config.agent_data_path, name), 'r')
+            data = json_lib.parse(fp.read())
+            fp.close()
+
+            data['time'] -= self._config.max_allowed_checkpoint_age + 1
+
+            fp = open(os.path.join(self._config.agent_data_path, name), 'w')
+            fp.write(json_lib.serialize(data))
+            fp.close()
+
+        # shift time on checkpoint files in the past, to make them stale.
+        shift_time_in_checkpoint_file('checkpoints.json')
+        shift_time_in_checkpoint_file('active-checkpoints.json')
+
+        # start manager.
```

Comment:
Add a comment here that we are specifically looking to make sure that the lines fifth and sixth do not show up since they were skipped over due to staleness.

Reply:
Done.